src/parse/validator.rs

```diff
@@ -250,131 +257,136 @@ impl<'help, 'app, 'parser> Validator<'help, 'app, 'parser> {
+                    self.p.app.color(),
+                ))
+            })
     }
 
     // Gathers potential conflicts based on used argument, but without considering requirements
     // and such
     fn gather_conflicts(&mut self, matcher: &mut ArgMatcher) {
         debug!("Validator::gather_conflicts");
-        for name in matcher.arg_names() {
-            debug!("Validator::gather_conflicts:iter: id={:?}", name);
-            // if arg is "present" only because it got default value
-            // it doesn't conflict with anything
-            //
-            // TODO: @refactor Do it in a more elegant way
-            if matcher
-                .get(name)
-                .map_or(false, |a| a.ty == ValueType::DefaultValue)
-            {
-                debug!("Validator::gather_conflicts:iter: This is default value, skipping.",);
-                continue;
-            }
-
-            if let Some(arg) = self.p.app.find(name) {
-                // Since an arg was used, every arg it conflicts with is added to the conflicts
-                for conf in &arg.blacklist {
-                    if self.p.app.find(conf).is_some() {
-                        if conf != name {
-                            self.c.insert(conf.clone());
+        matcher
+            .arg_names()
+            .filter(|name| {
+                debug!("Validator::gather_conflicts:iter: id={:?}", name);
+                // if arg is "present" only because it got default value
+                // it doesn't conflict with anything and should be skipped
+                let skip = matcher
+                    .get(name)
+                    .map_or(false, |a| a.ty == ValueType::DefaultValue);
+                if skip {
+                    debug!("Validator::gather_conflicts:iter: This is default value, skipping.",);
+                }
+                !skip
+            })
+            .for_each(|name| {
+                if let Some(arg) = self.p.app.find(name) {
+                    // Since an arg was used, every arg it conflicts with is added to the conflicts
+                    for conf in &arg.blacklist {
+                        /*
+                        for g_arg in self.p.app.unroll_args_in_group(&g.name) {
+                            if &g_arg != name {
+                                self.c.insert(g.id.clone()); // TODO ERROR is here - groups allow one arg but this line disallows all group args
+                            }
                         }
-                    } else {
-                        // for g_arg in self.p.app.unroll_args_in_group(conf) {
-                        //     if &g_arg != name {
-                        self.c.insert(conf.clone()); // TODO ERROR is here - groups allow one arg but this line disallows all group args
-                                                     //     }
-                                                     // }
+                        */
+                        self.c.insert(conf.clone());
```

Comment:
@rami3l any details on what this error was?  I think this is what is responsible for #3197 and I'd like to avoid any pitfalls that were previously uncovered.

Reply:
@epage Sorry, but git blame might be wrong this time, as I just moved those comments at the time, and I was also quite confused about them... Maybe you can track further down?

PS: I'm on vacation right now and I'll be happy to help with this a bit later if necessary :)

Reply:
I'm all for focusing on vacation!

Oh, when looking at this, I missed that that the two halves of the if/else were the same, that you just unified them.

Reply:
Closed thanks to https://github.com/clap-rs/clap/issues/3197.